core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +724,199 @@ oppia.config(['$provide', function($provide) {
+        init: function(editor) {
+          // Create the widget itself.
+          editor.widgets.add(ckName, {
+            button: componentDefn.tooltip,
+            inline: inlineCapable.indexOf(componentDefn.name) !== -1,
+            template: `<${tagName}></${tagName}>`,
+            edit: function(event) {
+              // Prevent default action since we are using our own edit modal.
+              event.cancel();
+              // Save this for creating the widget later.
+              var container = this.wrapper.getParent(true);
+              var that = this;
+              var customizationArgs = {};
+              customizationArgSpecs.forEach(function(spec) {
+                customizationArgs[spec.name] = that.data[spec.name] ||
+                                               spec.default_value;
+              });
+
+              rteHelperService._openCustomizationModal(
+                customizationArgSpecs,
+                customizationArgs,
+                function(customizationArgsDict) {
+                  for (var arg in customizationArgsDict) {
+                    if (customizationArgsDict.hasOwnProperty(arg)) {
+                      that.setData(arg, customizationArgsDict[arg]);
+                    }
+                  }
+                  // Actually create the widget.
+                  editor.widgets.finalizeCreation(container);
+                  $rootScope.$broadcast('render-rte-components');
+                },
+                function() {},
+                function() {});
+            },
+            upcast: function(element) {
+              return element.name === tagName;
+            },
+            data: function() {
+              var that = this;
+              customizationArgSpecs.forEach(function(spec) {
+                if (that.data[spec.name]) {
+                  that.element.setAttribute(
+                    spec.name + '-with-value',
+                    oppiaHtmlEscaper.objToEscapedJson(that.data[spec.name]));
+                }
+              });
+            },
+            init: function() {
+              var that = this;
+              customizationArgSpecs.forEach(function(spec) {
+                var value = that.element.getAttribute(
+                  spec.name + '-with-value');
+                if (value) {
+                  that.setData(
+                    spec.name, oppiaHtmlEscaper.escapedJsonToObj(value));
+                }
+              });
+              $rootScope.$broadcast('render-rte-components');
+            }
+          });
+        }
+      });
+    });
+  }
+]);
+
+oppia.directive('ckEditorRte', [
+  '$compile',
+  function($compile) {
+    return {
+      restrict: 'E',
+      scope: {
+        uiConfig: '&'
+      },
+      template: '<div contenteditable="true" style="outline: 1px solid gray">' +
+                '</div>',
+      require: '?ngModel',
+      link: function(scope, el, attr, ngModel) {
+        var widgets = [
```

Comment:
richTextComponents; please also rename below. I think we should only use 'widgets' when absolutely necessary, i.e. when mandated by CKEditor.